### DISCUSSION TITLE
How to instrument elasticsearch client 7.0 using OpenTelemetry

### DISCUSSION THREAD

**Discussion by venkatesh62891** [open-telemetry/opentelemetry-java-instrumentation#8867] [Q&A]
Hi Team,
I'm trying to manually instrument Elasticsearch client using OTEL. However, I see only java agent projects present in instrumentation folder but not any library folder which can  be used in application and not through javaagent. Kindly let me know how to go about this requirement, appreciate your time.

**Comment by mateuszrzeszutek** [comment]
Hey @venkatesh62891 ,

You can start by taking a look at what [the javaagent instrumentation is doing](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/elasticsearch/elasticsearch-rest-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/elasticsearch/rest/v7_0/RestClientInstrumentation.java) and trying to replicate it in your own code.
We don't have any library instrumentation for ES client because it does not expose any listener/interceptor types that would allow us to inject the instrumentation code into the client's behavior; which is why we have to do the bytecode instrumentation.

**Reply by trask** [reply]
cc a couple of folks at Elastic for visibility @AlexanderWert @estolfo @jackshirazi

**Reply by AlexanderWert** [reply]
Hi @venkatesh62891 ,

as @mateuszrzeszutek wrote above, unfortunately there's no easy and clean way of providing a library-level ES instrumentation (without Bytecode instrumentation). Is there some specific information you'd like to capture?

- ES calls are going through HTTP, so there always should be HTTP spans for ES requests. 
- If you need a custom ES span with additional information (but cannot use bytecode instrumentation), one workaround could be to wrap the `RestClient` into a custom class that would contain custom OTel instrumentation for the `performRequest` / `performAsyncRequest` methods of the `RestClient`. (However, this is rather hacky as the RestClient cannot be extended because of a private (default visibility) Constructor).
- We are currently working on native OTel instrumentation for the ES client, however, this will be only available for new client versions (probably 8.10+). With that, the instrumentation will be built directly into the client library and won't require bytecode instrumentation anymore.

**Reply by laurit** [reply]
Package private constructor isn't too big of an issue. You can place your wrapper in the same package as `RestClient`. Probably the best way would be to us a bytecode proxy and for this generated class is doesn't matter in which package it is in. At first glance adding library instrumentation for `RestClient` doesn't seem to be particularly difficult.

**Reply by venkatesh62891** [reply]
@laurit , @AlexanderWert @mateuszrzeszutek 
Thank you all for sharing your insights. If it helps, I’m currently using Elasticsearch client version 8.7.0. I’ve tried to add elasticsearch instrumentation javaagent jar as dependency but for some reason the bytecode instrumentation never gets registered. Is there a way to initiate the bytecode instrumentation through my springboot application?

@AlexanderWert 
I did see that RestClient internally uses HttpAsyncClient for making network calls, however there isn’t a library instrumentation available for the same. Currently only HttpClient 4.3 is supported with an instrumentation library. Please correct me if I’m wrong.

I did not understand the concept to wrap RestClient and instrument the wrapper. If it is to extend RestClient and use the extended version to instrument which means, start and end otel spans around super.performRequest() in the middle. This will still be a custom instrumentation where I have to capture same attributes as elasticsearch javaagent?  Am I thinking in right direction?

**Reply by mateuszrzeszutek** [reply]
> I’ve tried to add elasticsearch instrumentation javaagent jar as dependency but for some reason the bytecode instrumentation never gets registered. Is there a way to initiate the bytecode instrumentation through my springboot application?

The javaagent must be added through the `-javaagent` parameter to the JVM ([docs](https://opentelemetry.io/docs/instrumentation/java/automatic/#setup)); it's not a "normal" dependency that you just add to your classpath.

**Reply by venkatesh62891** [reply]
@mateuszrzeszutek I have tried the option to extend RestClient using the same package but still RestClientBuilder is another challenge which will create instance of only RestClient. Both RestClient and builder are pieced together and restricted to find a way to override the behavior.
**Solution-1**
1. Create wrapper for RestClient and extend RestClient in same package as RestClient.
2. Create a wrapper for RestClientBuilder and modify code to return RestClientWrapper instead of RestClient(RestClientBuilder need not and cannot be extended as it is final).
3. Build RestClientWrapper using RestClientBuilderWrapper and inject into ElasticSearchClient.
4. Do create and end span around super.performRequest(Request) with required span parameters same as java agent for Elasticsearch.

**Solution-2**
Since I'm using a spring boot application, I'm using AspectJAutoProxy
1. Create an Aspect around the method RestClient.performRequest.
2. Do create and end span around joinpoint.proceed with required span parameters same as java agent for Elasticsearch.

**Both will work just fine but be cautious with Solution-1.** Since the RestClientBuilder cannot be extended, a wrapper for builder has to added to the code alternate to RestClientBuilder class. Original Builder class can be modified by SDK providers and we may always have to sequence the update with every release version.

**Reply by mateuszrzeszutek** [reply]
> Since I'm using a spring boot application, I'm using AspectJAutoProxy

That's probably the better choice here 👍 
Using AspectJ, in this case you should be able to achieve very similar results as with the javaagent.

**Reply by laurit** [reply]
I created a pr with library instrumentation for elasticsearch rest client https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/8911

**Reply by venkatesh62891** [reply]
Hi @laurit , thank you for taking time to create the solution for Elasticsearch instrumentation. When can we expect the PR to be merged?

**Reply by trask** [reply]
Likely before the August release

**Comment by venkatesh62891** [comment]
> * We are currently working on native OTel instrumentation for the ES client, however, this will be only available for new client versions (probably 8.10+). With that, the instrumentation will be built directly into the client library and won't require bytecode instrumentation anymore.

@AlexanderWert Currently I'm using Elasticsearch client 8.7. When is the native client instrumentation be expected to hit the market?

**Reply by AlexanderWert** [reply]
Hi @venkatesh62891 , 
sorry for the delayed reply.

We are currently working on that, so it should be available with one of the next releases of the client (NOT before version 8.10).
